frame/two-phase-election-provider/src/macros.rs

```diff
@@ -0,0 +1,165 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Some helper functions/macros for this crate.
+
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		frame_support::debug::$level!(
+			target: $crate::LOG_TARGET,
+			concat!("🏦 ", $patter) $(, $values)*
+		)
+	};
+}
+
+#[deprecated = "May only be used for benchmarking purposes."]
+#[macro_export]
+macro_rules! voter_index_fn_linear {
+	(let $name:ident, $voters:ident, $t:ident) => {
+		let $name = |who: &<$t as frame_system::Config>::AccountId| -> Option<$crate::CompactVoterIndexOf<$t>> {
+			$voters
+				.iter()
+				.position(|(x, _, _)| x == who)
+				.and_then(|i| <usize as $crate::sp_std::convert::TryInto<$crate::CompactVoterIndexOf<$t>>>::try_into(i).ok())
+		};
+	};
+}
+
+#[macro_export]
+macro_rules! voter_index_generate_cache {
+	($name:ident, $voters:ident, $t:ident) => {
+		let mut $name:
+			$crate::sp_std::collections::btree_map::BTreeMap<
+				<$t as frame_system::Config>::AccountId,
+				usize
+			>
+			= $crate::sp_std::collections::btree_map::BTreeMap::new();
+		$voters.iter().enumerate().for_each(|(i, (x, _, _))| {
+			let _existed = $name.insert(x.clone(), i);
+			// if a duplicate exists, we only consider the last one. Defensive only, should never happen.
+			debug_assert!(_existed.is_none());
+		});
+	};
+}
+
+#[macro_export]
+macro_rules! voter_index_fn {
+	(let $name:ident, $voters:ident, $t:ident) => {
+		voter_index_generate_cache!(voters_map, $voters, $t);
+		let $name = |who: &<$t as frame_system::Config>::AccountId| -> Option<CompactVoterIndexOf<$t>> {
+			voters_map
+				.get(who)
+				.and_then(|i| <usize as $crate::sp_std::convert::TryInto<CompactVoterIndexOf<$t>>>::try_into(*i).ok())
+		};
+	};
+}
+
+#[macro_export]
+macro_rules! voter_index_fn_usize {
+	(let $name:ident, $voters:ident, $t:ident) => {
+		voter_index_generate_cache!(voters_map, $voters, $t);
+						let $name = |who: &<$t as frame_system::Config>::AccountId| -> Option<usize> {
+							voters_map.get(who).cloned()
+						};
+	};
+}
+
+#[macro_export]
+macro_rules! target_index_fn {
+	(let $name: ident, $targets:ident, $t:ident) => {
+		let $name = |who: &<$t as frame_system::Config>::AccountId| -> Option<$crate::CompactTargetIndexOf<$t>> {
+			$targets
+				.iter()
+				.position(|x| x == who)
+				.and_then(|i|
+					<
+						usize
+						as
+						$crate::sp_std::convert::TryInto<$crate::CompactTargetIndexOf<$t>>
+					>::try_into(i).ok()
+				)
+		};
+	};
+}
+
+#[macro_export]
+macro_rules! voter_at_fn {
+	(let $name:ident, $snap:ident, $t:ident) => {
+		let $name = |i: $crate::CompactVoterIndexOf<$t>| -> Option<<$t as frame_system::Config>::AccountId> {
+			<$crate::CompactVoterIndexOf<$t> as $crate::sp_std::convert::TryInto<usize>>::try_into(i)
+				.ok()
+				.and_then(|i| $snap
+					.get(i)
+					.map(|(x, _, _)| x)
+					.cloned()
+				)
+		};
+	};
+}
```

Comment:
same as Shawn, I would also prefer to write it as a function, you can do like this:
```rust
pub fn voter_at_fn<'a, T: crate::Config>(snap: &'a Vec<(T::AccountId, crate::VoteWeight, Vec<T::AccountId>)>) -> impl Fn(crate::CompactVoterIndexOf<T>) -> Option<<T as frame_system::Config>::AccountId> + 'a
	where
		crate::ExtendedBalance: From<crate::InnerOf<crate::CompactAccuracyOf<T>>>,
		crate::ExtendedBalance: From<crate::InnerOf<crate::OnChainAccuracyOf<T>>>,
{
	move |i: crate::CompactVoterIndexOf<T>| -> Option<<T as frame_system::Config>::AccountId> {
		<crate::CompactVoterIndexOf<T> as crate::sp_std::convert::TryInto<usize>>::try_into(i)
			.ok()
			.and_then(|i| snap.get(i).map(|(x, _, _)| x).cloned())
	}
}
```

Reply:
Done, last minute I had to switch to `Box` because with `impl Fn` you can't turbo fish and some types would fail to resolve.

Reply:
what do you mean ? the code above was compiling on my machine and it worked so far, do you say it doesn't work ?

Reply:
I see your commit now, you can still remove Box I think
[diff.txt](https://github.com/paritytech/substrate/files/5814007/diff.txt)

Reply:
Did you compile all the tests as well? It failed for me when compiling only a bunch of tests. I will push an experimental commit for you to see, maybe I missed sth.

Reply:
I get the same error as I had before with your patch when compiling the tests. 

```
error[E0277]: expected a `Fn<(&'r u64,)>` closure, found `impl for<'r> Fn<(&'r <mock::Runtime as frame_system::Config>::AccountId,)>`
```

Reply:
Seems kinda strange as I expected the compiler to be able to resolve this. Even when you turbo-fish the call to `assignment_ratio_to_staked_normalized` which is raising the issue like `assignment_ratio_to_staked_normalized::<<Runtime as system::Config>::AccountId>` won't help. 

Either way I decided not to pursue it as they are hardly distinguishable. Happy to hear a solution if you have any.

Reply:
Yes I run test `cargo test -p pallet-two-phase-election-provider` I guess the compiler is failing, but not my nightly.
Then it is fine with Box